### PR TITLE
Reorganise test mod into packages

### PR DESCRIPTION
**Pull request by marvin-roesch**
**TL/DR: Things should change, only contributors need to (partially) care.**

# Motivation
A quick glance at the [current state](https://github.com/MinecraftForge/MinecraftForge/tree/3b580f643e6838b9f94f01fd98cbe5ed70c63fd1/src/test/java/net/minecraftforge/debug) of the debug mod package is quite telling. Finding any test mod to use as an implementation example is quite hard, both due to the lack of any package structure as well as wildly varying names. This PR aims to mitigate these problems by proposing a standardised structure and guidelines for consistent naming. Further improvements are also mentioned.

# Proposal
I'd like to introduce the core ideas behind this pull request and will outline a way of implementing them properly.

## Package Structure
Test mods are supposed to serve as minimal examples that show off as much of a new feature as possible. Accordingly, they should be easy to find and understand. A *clear* package structure goes a long way to help with this, especially if it fulfils the following criteria:

 - **Semantic:** Package names should be *meaningful*. This implies that you should be able to tell at a glance what the package contains. For test mods, this can be achieved by simply putting a (new) mod in the package corresponding to the category the feature most closely relates to.
  *If no fitting package exists yet, a new one should be created if it doesn't violate any other criteria.*
 - **Easy to navigate:** This criterion has two dimensions:
To ensure it, package names need to be *concise* (i.e. *short and meaningful*), allowing you to quickly scan through the list of packages to find the one most closely fitting your needs.
Moreover, packages should be *small* as far as class count is concerned. The main issue the current setup is suffering from is that there's a lot of classes in a single package, finding the mod you're looking for is impossible. Keeping packages small (excluding subpackages) allows for faster access to the correct class.
 - **Hierarchical:** Top-level categories should be very *broad*, encompassing a large, general category in the game, such as 'blocks' or 'entities'. The further down you go the hierarchy, the more *fine-grained* categories should become, always describing some subset of the parent.
However, the hierarchy should also be kept fairly *shallow*, not exceeding a depth of 3 nested packages.<br> *If you go any deeper, chances are that a category is too specific to a given feature and would not be reusable by others. Opt for the best-fitting higher level category in such cases.*

These criteria appear to be self-evident, but they must still be consistently adhered to.

## Class/Mod Naming
The current names of all the test mods follows a wild mix of styles. The following is a set of rules to follow in order to get a canonical and consistent class naming scheme:
 1. Mod class names *always* end in the `Test` suffix.
 2. If the mod demonstrates a specific event, the class name has to have the *exact* class name of the event as prefix as well as an `Event` suffix (if the class name does not contain it). If the event is an inner class of another, that class's prefix (i.e. everything before `Event`) is used as further prefix. If words are duplicated, they may be omitted as long as no ambiguity is generated. The same applies if the context can be derived from the package.<br>
*Examples*: `LivingKnockBackEventTest`, `CheckSpawnEvent` (both category 'Living' as well as duplicated 'Spawn' omitted), `BiomeDecorateEventTest` (Inner class, first 'Decorate' omitted)
 3. If the mod directly relates to a specific class, that class's name is used as a prefix. If it relates to any member of that class or a feature local to it, that member's/feature's name is used before the `Test` suffix. As with rule 2, any duplicated words as well package-derivable context may be omitted.<br>
*Examples*: `VertexBufferNormalTest`, `OnItemUseFirstTest` (method from class `Item`, mod in `items` package)
 4. If the mod serves as an example for a (new) system, feature or fix thereof, a *concise name for the feature is used as prefix. If it demonstrates a subset of a bigger feature, up to *3 more* describing words may be added after it.<br>
*Examples*: `BigNetworkMessageTest`, `HorseArmorTest`
 5. If the mod tests a specific annotation (e.g. some sort of automatic injection etc.), the name should be formed according to rule (4) but with an additional `Annotation` suffix after the main class name.<br>
*Examples*: `ConfigAnnotationTest`, `ObjectHolderAnnotationTest`
 6. If none of the above rules apply, a *sensible* and concise way should be chosen, describing the purpose of the test mod uniquely and easily searchable.

The name of the test mod should follow the same scheme and must match the class name except for a few transformations. The 'Test' suffix always is separated by a comma and every non-class part (i.e. not the name of an `Event` or other class) of the class name also becomes a separate word. The name should follow title-casing.

## Test Mod Structure, Best Practices
*Note: This section still is very much work-in-progress, as inconsistencies need to be identified and a standard worked out.*

If a test mod wants to serve as an example for modders, it should be well structured and following the current best practices. Furthermore, readers should be able to navigate from one test mod to another without having to deal with completely different code in each mod. Some efforts in this direction have already been made, but especially the larger test mods vary a lot in structure.

## Documentation, Code Style
Most test mods have lacking or no documentation at all. We do no not need complete JavaDocs or a detailed document, but *all non-trivial* parts of the test mod (especially those related to the feature demonstrated, obviously) should be sufficiently and understandably commented. If the pull request contains a more detailed description that would not fit comments, it may be referenced (by number only) in the JavaDoc comment for the class.

Another factor in understanding example mods is a consistent, clean code style (design-wise, not formatting). Of course every contributor has slightly different preferences, but should put in effort to keep in line with the existing tests. Most importantly, even if dealing with test mods, no quick workarounds or 'hacks' should be employed just because it's faster to get things done. Additionally, any prototyping/commented out code should be cleaned up or removed before opening a pull request.

# Implementation
I've completed the following stages of the above proposal so far:

 - [x] Restructure classes into semantic packages
 - [x] Adopt consistent naming guidelines
 - Restructure mod code in a consistent manner (later)
 - Document test mods succinctly (later as well)

I would appreciate some feedback before proceeding beyond the class renaming stage as I want to be confident to go forward with the proposed (and already implemented) package structure and naming scheme. Especially the `mods` (for anything not directly related to the game but mods only) and `gameplay` (things like recipes, brewing and enchanting) packages I'm not quite satisfied with yet.

# Prospect
In the long run, I'd like to revamp test mods (or rather loading them) altogether. I've been thinking about writing a 'coremod' (it'd functionally be one, but shouldn't need to do any ASM) that serves as a dedicated test mod loader. Test mods would then use a `@TestMod` annotation that's a simpler variant of the current `@Mod` one which only allows for dependencies and some simple client/server related things. Mod IDs could potentially also be auto-generated by the loader, most likely in a format akin to `ft_<name>` (`ft` signifying 'Forge Test').

The test mod loader would add a config and corresponding GUI (either in MC or through Swing/JavaFX) that allows selecting which test mods need to be loaded at all. This way, we effectively get disabling for free and could even automate startup of test mods on Travis. The (simple) dependency system would give us the ability to write a generic set of mods adding blocks, items, entities and dimension(s) to perform testing on, without polluting the game with unnecessary stuff.

### CODE REVIEW

**Comment by mezz** [comment]
TLDR Paleo organized the test mods some more so they're not the wild west, and proposes some more standards for test mods.

I don't think we should make test mods very different from regular mods. Part of the benefit of having them this way is that they run the same code that normal mods do. If you want to enhance it, I think it should be an enhancement for all mods if possible.

**Comment by jabelar** [comment]
I think that the organization is possibly a bit contrived. Like any classification scheme we'll end up with things that don't fit neatly. For example, a test mod for a PR that is focused on block events for fluids -- is that a fluid category, or an event category, or a block category? Is a feature semantically similar to a bug fix or a clean up? And personally I'd probably make Events a category as I feel that just like in the actual Forge package scheme that events have more in common with each other than the specific thing the event is about. But of course that is arguable. But your scheme is worth a try I guess.

I like a lot of the other suggestions though (especially an eventual loader mechanism to allow you to only load the test mod of choice). 

In particular I think maybe the documentation style/template should be enforced -- you could come up with a comment/javadoc format that includes required information such as the PR number, how to confirm the workings in game, how to modify the code to check that the mechanism of the new feature/fix works, and preferably somehow highlight the specific portion of the test mod which is unique to the PR.

**Comment by marvin-roesch** [comment]
@jabelar: Arguably any structure will have its benefits and downsides. My reasoning for the one proposed here is that it follows gameplay in some way. Lumping all events together, for example, might make sense from some perspective, but we'd have subcategories again that classify them by type of event (entity, rendering etc.). Hence, just going by these categories right away should still make finding the mod you seek easy.

Regarding your fluid block event example: I'd argue that fluids are a subsystem of their own, so they deserve their own category where this test mod would go. The 'block' category in my scheme would serve for features affecting all blocks or a single block that doesn't fit any other category.

@mezz: If you're okay with it, I'll update this PR this evening to accommodate any new test mods and then call it done and mergeable. What I'm going to call phase 2 (consistent code style and docs) is better suited for a separate PR and will take more time I currently don't have.

**Comment by jabelar** [comment]
Yeah, I'm actually okay with almost any scheme, just pointing out that we shouldn't get our hopes up that everything will be fully tidy afterwards as there will definitely be ones that defy easy categorization. But there are certainly lots that can fit neatly so let's give it a try.

**Comment by marvin-roesch** [comment]
This is by no means stale, I'll see to new test mods being covered as well sometime this week.

**Comment by mezz** [comment]
Let me know when you update it and I can take a look.

**Comment by marvin-roesch** [comment]
I think this is ready now, @mezz. Looks like there weren't any new test mods added since my last merge.

**Comment by mezz** [comment]
Checked it out and looks good to me.
In the absolute worst case, this isn't breaking anything and we can refine it later, so I'll merge it now.

**Comment by bs2609** [comment]
```
com.google.gson.JsonSyntaxException: Could not find IIngredientFactory: net.minecraftforge.debug.CraftingSystemTest$IngredientFactory
```

**Comment by mezz** [comment]
Thanks, fixed that and the `asdkfaskld` issue in https://github.com/MinecraftForge/MinecraftForge/commit/cfe3867a70d4709b069e5e57d0d048982ae979a7
